be/src/exec/pipeline/exchange/local_exchange.cpp

```diff
@@ -174,6 +162,130 @@ Status PartitionExchanger::accept(const ChunkPtr& chunk, const int32_t sink_driv
+    LocalExchanger::close(state);
+}
+
+Status OrderedPartitionExchanger::accept(const ChunkPtr& chunk, const int32_t sink_driver_sequence) {
+    DCHECK_EQ(sink_driver_sequence, 0);
+
+    Columns partition_columns(_partition_exprs.size());
+    for (size_t i = 0; i < partition_columns.size(); ++i) {
+        ASSIGN_OR_RETURN(partition_columns[i], _partition_exprs[i]->evaluate(chunk.get()));
+        DCHECK(partition_columns[i] != nullptr);
+    }
+
+    ChunkPtr output_chunk = chunk;
+
+    size_t channel_id;
+    size_t min_channel_id = _find_min_channel_id();
+    if (_previous_chunk == nullptr || _previous_channel_id == min_channel_id) {
+        channel_id = min_channel_id;
+    } else {
+        // Check if the joint of two consecutive chunks are the same
+        bool is_joint_equal = true;
+        for (size_t i = 0; i < partition_columns.size(); ++i) {
+            auto cmp = partition_columns[i]->compare_at(0, _previous_chunk->num_rows() - 1, *partition_columns[i], 1);
+            if (cmp != 0) {
+                is_joint_equal = false;
+                break;
+            }
+        }
+
+        if (!is_joint_equal) {
+            // The first row of current chunk is the start of a new partition, so
+            // send the chunk to the channel with the minimum number of rows.
+            channel_id = min_channel_id;
+        } else {
+            bool is_current_of_same_partition = true;
+            for (auto& column : partition_columns) {
+                auto cmp = column->compare_at(0, column->size() - 1, *column, 1);
+                if (cmp != 0) {
+                    is_current_of_same_partition = false;
+                    break;
+                }
+            }
+            if (is_current_of_same_partition) {
+                channel_id = _previous_channel_id;
+            } else {
+                // Found partition end that belongs to the first row of current chunk, and split the chunk into two parts:
+                // 1. The first part is the rows of the same partition as the last row of previous chunk, and send it to previous channel
+                // 2. The second part is the rows of the different partition, and send it to the channel with the minimum number of rows.
+
+                auto _find_first_not_equal = [](Column* column, int64_t target, int64_t start, int64_t end) {
+                    while (start + 1 < end) {
+                        int64_t mid = start + (end - start) / 2;
+                        if (column->compare_at(target, mid, *column, 1) == 0) {
+                            start = mid;
+                        } else {
+                            end = mid;
+                        }
+                    }
+                    if (column->compare_at(target, end - 1, *column, 1) == 0) {
+                        return end;
+                    }
+                    return end - 1;
+                };
+
+                int64_t end = chunk->num_rows();
```

Comment:
use partition_boundary or boundary instead of end;

Reply:
it the initial boundary of binary search.